### DISCUSSION TITLE
Idea: support additional Prisma Schema fragments

### DISCUSSION THREAD

**Discussion by JedWatson** [keystonejs/keystone#6994] [Feature requests]
I've been thinking about how Keystone currently needs to "own" the whole Prisma schema, which means it "owns" the whole database (because of how we use Migrate) and how we might be able to relax that for people who want things (anything, really, additional tables, columns, indexes, etc) in their DB that isn't sourced from a Keystone list or field.

This has come up a few times for people who have an existing database and are trying to incrementally move to Keystone, or only have Keystone manage _part_ of that database. It would also be nice to have a way to do Prisma things that Keystone doesn't support in its schema.

It occurred to me that we could actually open up an escape hatch that makes this _really_ simple, by allowing people to write additional strings into the Prisma schema that we generate.

For example:

```ts
config({
  db: {
    provider: 'sqlite',
    url: process.env.DATABASE_URL || 'file:./keystone-example.db',
    addPrismaSchema: `
      model MyCustomThing {
        id          String    @id @default(cuid())
        title       String    @default("")
      
        @@index([title])
      }
    `
  },
  lists: {
    Person: list({
      fields: {
        name: text(),
        email: text(),
      },
      db: {
        addPrismaSchema: `
          my_custom_field String

          @@index([name, email])
        `
      },
    }),
  },
});
```

This would all get merged into a single Prisma schema by Keystone:

```
datasource sqlite {
  url      = env("DATABASE_URL")
  provider = "sqlite"
}

generator client {
  provider   = "prisma-client-js"
  output     = "node_modules/.prisma/client"
  engineType = "binary"
}

model Person {
  id                 String    @id @default(cuid())
  name               String    @default("")
  email              String    @default("")
  my_custom_field    String

  @@index([name, email])
}

model MyCustomThing {
  id       String    @id @default(cuid())
  title    String    @default("")

  @@index([title])
}
```

This would open up a few possibilities:

* Add tables that are available to the Prisma Client but not the Keystone API (but could be used e.g in custom queries and mutations, virtual fields, hooks etc)
* Add relations between those tables and Keystone List tables
* Add columns to Keystone List tables using Prisma features Keystone doesn't support
* Add other features Keystone doesn't expose yet like multi-column indexes to Keystone List tables 

There's an issue in Prisma for supporting multiple schema files which may make the DX of this nicer down the track, we should keep an eye on: prisma/prisma#2377

Also the [official workaround](https://www.prisma.io/docs/concepts/components/introspection#introspecting-only-a-subset-of-your-database-schema) for migration and introspection using a different database user with limited permissions is a thing... but using this "inject strings into your prisma schema" approach, if you want "invisible to Keystone" things in the database you still get the benefit of Prisma Migrate, and the Prisma Client, which is more what I'm going for here.

**Comment by Cauen** [comment]
I think its also great to enable integration with prisma [generators](https://www.prisma.io/docs/concepts/components/prisma-schema/generators), that will be necessary when the [nexus example](https://github.com/keystonejs/keystone/tree/main/examples/extend-graphql-schema-nexus) improves.
>There's also a Prisma plugin for Nexus in development here: https://github.com/prisma/nexus-prisma
When it's ready, it would make a good addition to this example (showing how to integrate the Prisma plugin with the Keystone-generated Prisma schema to auto-generate Nexus schema)

Once the [integration](https://nexus.prisma.io/usage) requires a generator:

```
generator client {
  provider = "prisma-client-js"
}

generator nexusPrisma {
   provider = "nexus-prisma"
}

/// This is a user!
model User {
  /// This is an id!
  id  String  @id
}
```

**Comment by molomby** [comment]
This has more or less been implemented as `db.extendPrismaSchema`, available on the root, list and field-level configs 🎉 See the [`extend-prisma-schema` example project](https://github.com/keystonejs/keystone/tree/main/examples/extend-prisma-schema) for details.
